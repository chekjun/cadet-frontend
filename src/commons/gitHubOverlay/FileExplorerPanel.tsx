import { Classes, InputGroup, ITreeNode, Tree } from '@blueprintjs/core';
import classNames from 'classnames';
import React, { Component } from 'react';

import { GitHubFileNodeData } from './GitHubFileNodeData';
import { GitHubTreeNodeCreator } from './GitHubTreeNodeCreator';

export interface IFileExplorerPanelProps {
  repoFiles: ITreeNode<GitHubFileNodeData>[];
  repoName: string;
  pickerType: string;
  filePath: string;
  setFilePath: any;
  setCommitMessage: any;
}

export interface IFileExplorerPanelState {
  repoFiles: ITreeNode<GitHubFileNodeData>[];
  commitMessage: string;
}

export class FileExplorerPanel extends Component<IFileExplorerPanelProps, IFileExplorerPanelState> {
  constructor(props: IFileExplorerPanelProps) {
    super(props);
    this.handleNodeClick = this.handleNodeClick.bind(this);
    this.handleNodeCollapse = this.handleNodeCollapse.bind(this);
    this.handleNodeExpand = this.handleNodeExpand.bind(this);
    this.forEachNode = this.forEachNode.bind(this);
    this.handleFileNameChange = this.handleFileNameChange.bind(this);
    this.handleCommitMessageChange = this.handleCommitMessageChange.bind(this);
  }

  public state: IFileExplorerPanelState = {
    repoFiles: this.props.repoFiles,
    commitMessage: ''
  };

  private handleNodeClick(
    treeNode: ITreeNode<GitHubFileNodeData>,
    _nodePath: number[],
    e: React.MouseEvent<HTMLElement>
  ) {
    const originallySelected = treeNode.isSelected;

    if (!e.shiftKey) {
      this.forEachNode(this.state.repoFiles, n => (n.isSelected = false));
    }

    treeNode.isSelected = originallySelected == null ? true : !originallySelected;
    const newFilePath = treeNode.nodeData !== undefined ? treeNode.nodeData.filePath : '';
    if (treeNode.isSelected) {
      this.props.setFilePath(newFilePath);
    } else {
      this.props.setFilePath('');
    }
    this.setState(this.state);
  }

  private handleNodeCollapse(treeNode: ITreeNode<GitHubFileNodeData>) {
    treeNode.isExpanded = false;
    this.setState(this.state);
  }

  private async handleNodeExpand(treeNode: ITreeNode<GitHubFileNodeData>) {
    treeNode.isExpanded = true;

    if (treeNode.nodeData !== undefined && !treeNode.nodeData.childrenRetrieved) {
      treeNode.childNodes = await GitHubTreeNodeCreator.getChildNodes(
        this.props.repoName,
        treeNode.nodeData.filePath
      );
      //treeNode.childNodes = await this.props.getChildNodes(treeNode.nodeData.filePath);
      treeNode.nodeData.childrenRetrieved = true;
    }

    this.setState(this.state);
  }

  private forEachNode(
    treeNodes: ITreeNode<GitHubFileNodeData>[],
    callback: (node: ITreeNode<GitHubFileNodeData>) => void
  ) {
    if (treeNodes == null) {
      return;
    }

    for (const node of treeNodes) {
      callback(node);
      if (node.childNodes !== undefined) {
        this.forEachNode(node.childNodes, callback);
      }
    }
  }

  render() {
    return (
      <div className={classNames(Classes.DIALOG_BODY, 'file-step')}>
        <Tree
          contents={this.state.repoFiles}
          onNodeClick={this.handleNodeClick}
          onNodeCollapse={this.handleNodeCollapse}
          onNodeExpand={this.handleNodeExpand}
          className={Classes.ELEVATION_0}
        />
        {this.props.pickerType === 'Save' && (
          <div>
            <InputGroup
              onChange={this.handleFileNameChange}
              placeholder={'Enter File Name'}
              value={this.props.filePath}
            />
            <InputGroup
              onChange={this.handleCommitMessageChange}
              placeholder={'Enter Commit Message'}
              value={this.state.commitMessage}
            />
          </div>
        )}
      </div>
    );
  }

  handleFileNameChange(e: any) {
    this.props.setFilePath(e.target.value);
  }

  handleCommitMessageChange(e: any) {
    this.setState({ commitMessage: e.target.value });
    this.props.setCommitMessage(e.target.value);
  }
}
