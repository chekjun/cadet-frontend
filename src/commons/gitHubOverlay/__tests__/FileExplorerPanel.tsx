import { act, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mount } from 'enzyme';

import { FileExplorerPanel } from '../FileExplorerPanel';

test('Test file/folder click calls setFilePath', () => {
  act(() => {
    render(<FileExplorerPanel {...props} />);
  });
  const leftClick = { button: 1 };
  fireEvent.click(screen.getByText('folder1'), leftClick);
  expect(props.setFilePath).toHaveBeenCalledTimes(1);
  fireEvent.click(screen.getByText('folder2'), leftClick);
  expect(props.setFilePath).toHaveBeenCalledTimes(2);
  fireEvent.click(screen.getByText('file3'), leftClick);
  expect(props.setFilePath).toHaveBeenCalledTimes(3);
});

test('Test file/folder clicked twice sets filepath to empty string', () => {
  act(() => {
    render(<FileExplorerPanel {...props} />);
  });
  const leftClick = { button: 1 };
  fireEvent.click(screen.getByText('file3'), leftClick);
  expect(props.setFilePath).toHaveBeenCalledTimes(1);
  fireEvent.click(screen.getByText('file3'), leftClick);
  expect(props.setFilePath).toHaveBeenCalledTimes(2);
  expect(props.filePath).toBe('');
});

test('Test file name input typed calls setFilePath', () => {
  act(() => {
    render(<FileExplorerPanel {...props} />);
  });
  userEvent.type(screen.getByPlaceholderText('Enter File Name'), 'a');
  expect(props.setFilePath).toHaveBeenCalledTimes(1);
  userEvent.type(screen.getByPlaceholderText('Enter File Name'), 'file3');
  expect(props.setFilePath).toHaveBeenCalledTimes(6);
});

test('Test commit message input typed calls setCommitMessage', () => {
  act(() => {
    render(<FileExplorerPanel {...props} />);
  });
  userEvent.type(screen.getByPlaceholderText('Enter Commit Message'), 'a');
  expect(props.setCommitMessage).toHaveBeenCalledTimes(1);
  userEvent.type(screen.getByPlaceholderText('Enter Commit Message'), 'message');
  expect(props.setCommitMessage).toHaveBeenCalledTimes(8);
});

test('Test file list renders correctly', () => {
  const FEP = mount(<FileExplorerPanel {...props} />);
  expect(FEP.debug()).toMatchSnapshot();
});

const repoFiles = [
  {
    id: 0,
    label: 'folder1',
    nodeData: {
      childrenRetrieved: true,
      filePath: 'folder1',
      fileType: 'dir',

      constructor(path: string, type: string) {
        this.filePath = path;
        this.fileType = type;
      }
    },
    childNodes: [
      {
        id: 3,
        label: 'file1',
        nodeData: {
          childrenRetrieved: false,
          filePath: 'folder1/file1',
          fileType: 'file',

          constructor(path: string, type: string) {
            this.filePath = path;
            this.fileType = type;
          }
        }
      },
      {
        id: 4,
        label: 'file2',
        nodeData: {
          childrenRetrieved: false,
          filePath: 'folder1/file2',
          fileType: 'file',

          constructor(path: string, type: string) {
            this.filePath = path;
            this.fileType = type;
          }
        }
      }
    ]
  },
  {
    id: 1,
    label: 'folder2',
    nodeData: {
      childrenRetrieved: false,
      filePath: 'folder2',
      fileType: 'dir',

      constructor(path: string, type: string) {
        this.filePath = path;
        this.fileType = type;
      }
    }
  },
  {
    id: 2,
    label: 'file3',
    nodeData: {
      childrenRetrieved: false,
      filePath: 'file3',
      fileType: 'file',

      constructor(path: string, type: string) {
        this.filePath = path;
        this.fileType = type;
      }
    }
  }
];

const props = {
  repoFiles: repoFiles,
  repoName: 'YEETREPO',
  pickerType: 'Save',
  filePath: '',
  setFilePath: jest.fn(),
  setCommitMessage: jest.fn()
};
