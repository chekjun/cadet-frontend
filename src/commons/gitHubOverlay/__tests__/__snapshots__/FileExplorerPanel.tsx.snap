// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test file list renders correctly 1`] = `
"<FileExplorerPanel repoFiles={{...}} repoName=\\"YEETREPO\\" pickerType=\\"Save\\" filePath=\\"\\" setFilePath={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} setCommitMessage={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
  <div className=\\"bp3-dialog-body\\">
    <Blueprint3.Tree contents={{...}} onNodeClick={[Function: bound handleNodeClick]} onNodeCollapse={[Function: bound handleNodeCollapse]} onNodeExpand={[Function: bound handleNodeExpand]} className=\\"bp3-elevation-0 FilePanel\\">
      <div className=\\"bp3-tree bp3-elevation-0 FilePanel\\">
        <ul className=\\"bp3-tree-node-list bp3-tree-root\\">
          <Blueprint3.TreeNode id={0} label=\\"folder1\\" nodeData={{...}} childNodes={{...}} isSelected={false} contentRef={[Function (anonymous)]} depth={0} onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onCollapse={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onExpand={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} path={{...}}>
            <li className=\\"bp3-tree-node\\">
              <div className=\\"bp3-tree-node-content bp3-tree-node-content-0\\" onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]}>
                <Blueprint3.Icon className=\\"bp3-tree-node-caret bp3-tree-node-caret-closed\\" onClick={[Function (anonymous)]} icon=\\"chevron-right\\">
                  <span onClick={[Function (anonymous)]} icon=\\"chevron-right\\" className=\\"bp3-icon bp3-icon-chevron-right bp3-tree-node-caret bp3-tree-node-caret-closed\\" title={[undefined]}>
                    <svg fill={[undefined]} data-icon=\\"chevron-right\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
                      <desc>
                        chevron-right
                      </desc>
                      <path d=\\"M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z\\" fillRule=\\"evenodd\\" />
                    </svg>
                  </span>
                </Blueprint3.Icon>
                <Blueprint3.Icon className=\\"bp3-tree-node-icon\\" icon={[undefined]} />
                <span className=\\"bp3-tree-node-label\\">
                  folder1
                </span>
              </div>
              <Blueprint3.Collapse isOpen={false} component=\\"div\\" keepChildrenMounted={false} transitionDuration={200}>
                <div className=\\"bp3-collapse\\" style={{...}}>
                  <div className=\\"bp3-collapse-body\\" style={{...}} aria-hidden={false} />
                </div>
              </Blueprint3.Collapse>
            </li>
          </Blueprint3.TreeNode>
          <Blueprint3.TreeNode id={1} label=\\"folder2\\" nodeData={{...}} isSelected={false} contentRef={[Function (anonymous)]} depth={0} onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onCollapse={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onExpand={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} path={{...}}>
            <li className=\\"bp3-tree-node\\">
              <div className=\\"bp3-tree-node-content bp3-tree-node-content-0\\" onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]}>
                <span className=\\"bp3-tree-node-caret-none\\" />
                <Blueprint3.Icon className=\\"bp3-tree-node-icon\\" icon={[undefined]} />
                <span className=\\"bp3-tree-node-label\\">
                  folder2
                </span>
              </div>
              <Blueprint3.Collapse isOpen={false} component=\\"div\\" keepChildrenMounted={false} transitionDuration={200}>
                <div className=\\"bp3-collapse\\" style={{...}}>
                  <div className=\\"bp3-collapse-body\\" style={{...}} aria-hidden={false} />
                </div>
              </Blueprint3.Collapse>
            </li>
          </Blueprint3.TreeNode>
          <Blueprint3.TreeNode id={2} label=\\"file3\\" nodeData={{...}} isSelected={true} contentRef={[Function (anonymous)]} depth={0} onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onCollapse={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onExpand={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} path={{...}}>
            <li className=\\"bp3-tree-node bp3-tree-node-selected\\">
              <div className=\\"bp3-tree-node-content bp3-tree-node-content-0\\" onClick={[Function (anonymous)]} onContextMenu={[Function (anonymous)]} onDoubleClick={[Function (anonymous)]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]}>
                <span className=\\"bp3-tree-node-caret-none\\" />
                <Blueprint3.Icon className=\\"bp3-tree-node-icon\\" icon={[undefined]} />
                <span className=\\"bp3-tree-node-label\\">
                  file3
                </span>
              </div>
              <Blueprint3.Collapse isOpen={false} component=\\"div\\" keepChildrenMounted={false} transitionDuration={200}>
                <div className=\\"bp3-collapse\\" style={{...}}>
                  <div className=\\"bp3-collapse-body\\" style={{...}} aria-hidden={false} />
                </div>
              </Blueprint3.Collapse>
            </li>
          </Blueprint3.TreeNode>
        </ul>
      </div>
    </Blueprint3.Tree>
    <div>
      <Blueprint3.InputGroup onChange={[Function: bound handleFileNameChange]} placeholder=\\"Enter File Name\\" value=\\"\\">
        <div className=\\"bp3-input-group\\">
          <input type=\\"text\\" onChange={[Function: bound handleFileNameChange]} placeholder=\\"Enter File Name\\" value=\\"\\" className=\\"bp3-input\\" style={{...}} />
        </div>
      </Blueprint3.InputGroup>
      <Blueprint3.InputGroup onChange={[Function: bound handleCommitMessageChange]} placeholder=\\"Enter Commit Message\\" value=\\"\\">
        <div className=\\"bp3-input-group\\">
          <input type=\\"text\\" onChange={[Function: bound handleCommitMessageChange]} placeholder=\\"Enter Commit Message\\" value=\\"\\" className=\\"bp3-input\\" style={{...}} />
        </div>
      </Blueprint3.InputGroup>
    </div>
  </div>
</FileExplorerPanel>"
`;
